src/meta/server_state.cpp

```diff
@@ -3269,8 +3285,321 @@ void server_state::set_max_replica_count(configuration_set_max_replica_count_rpc
              response.old_max_replica_count,
              new_max_replica_count);
 
-    // TODO: update partition-level and app-level max_replica_count
-    response.err = ERR_OK;
+    do_update_max_replica_count(app, app_id, partition_count, rpc);
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void server_state::do_update_max_replica_count(std::shared_ptr<app_state> &app,
+                                               int32_t app_id,
+                                               int32_t partition_count,
+                                               configuration_set_max_replica_count_rpc rpc)
+{
+    std::shared_ptr<std::vector<error_code>> results(new std::vector<error_code>(partition_count));
+    std::shared_ptr<std::atomic_int> updates_in_progress(new std::atomic_int(partition_count));
```

Comment:
> Why would you define `results` and `updates_in_progress` as shared_ptr?

Because both of them will be dispatched as callback into `update_partition_max_replica_count` of each partition, to be gathered at last.

Reply:
You can use the following variety called `partitions_in_progress`, no need to define a shared_ptr called `updates_in_progress`.
https://github.com/XiaoMi/rdsn/blob/b8942ecdd775dd3b22cb67941453ba812c847bc7/src/meta/meta_data.h#L308

As the `results` store the each partition update error code, would you like to do more error handler if any of it is not error_ok?

Reply:
> You can use the following variety called `partitions_in_progress`, no need to define a shared_ptr called `updates_in_progress`.
> 
> https://github.com/XiaoMi/rdsn/blob/b8942ecdd775dd3b22cb67941453ba812c847bc7/src/meta/meta_data.h#L308
> 
> As the `results` store the each partition update error code, would you like to do more error handler if any of it is not error_ok?

Yes, this is a synchronization to gather all results, then decide how to process these errors if any, or do response to client though currently dassert(false) is executed if there is any error.